### PR TITLE
Fix SlayerPlugin NPEs

### PR DESCRIPTION
**Pull request by AWarbear**
Fixes NPE caused from getting Otherworldly beings as a task due to it not being found from the task list.

### CODE REVIEW

**Comment by AWarbear** [comment]
I'll fix the cause of the NPE so that it will stop throwing NPEs every game tick when it doesn't know the task when future slayer monsters are released. Will also add all the other missing slayer tasks that have been reported missing in issues.
So don't merge yet.

**Comment by Adam-** [comment]
Is there any way we can not call buildTargetsToHighlight per tick?

**Comment by AWarbear** [comment]
There should actually be no reason to build the target (name) list anytime other than when the task name changes.

It seems that its updated per tick to stay accurate with the npcs that are currently in view. Is there a way to be notified when an npc becomes visible to the client?

**Comment by Joshua-F** [comment]
> It seems that its updated per tick to stay accurate with the npcs that are currently in view. Is there a way to be notified when an npc becomes visible to the client?

[runelite-api/src/main/java/net/runelite/api/events/NpcSpawned.java](runelite-api/src/main/java/net/runelite/api/events/NpcSpawned.java)
[runelite-api/src/main/java/net/runelite/api/events/NpcDespawned.java](runelite-api/src/main/java/net/runelite/api/events/NpcDespawned.java)

**Comment by AWarbear** [comment]
The 3rd commit fixes #2295

That should be all the missing tasks that have been reported. I'll move the highlight updating logic away from the game tick tomorrow.

**Comment by Nightfirecat** [comment]
@AWarbear I'd recommend you not title your commit "Partially fixes ..." as it matches [GitHub's keyword criteria](https://help.github.com/articles/closing-issues-using-keywords/) and would register as "fixes ...". Maybe change the wording to "Ref: ..." instead?

**Comment by AWarbear** [comment]
This should now be ready to be checked over and merged

Regarding Issue #1800 Flesh crawlers work fine as also stated in PR #1906 that has been open for a while to add baby dragons to dragon tasks. Should I add the baby dragons to this PR as it won't merge very well with the edited SlayerPlugin class?

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 1]
```diff
@@ -27,16 +27,15 @@
 
 import com.google.common.eventbus.Subscribe;
 import com.google.inject.Provides;
+
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: These new lines shouldn't be here https://github.com/runelite/runelite/wiki/Code-Conventions#imports
As well as wildcard imports and other import issues.

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 2]
```diff
@@ -67,9 +64,9 @@
 import net.runelite.client.util.Text;
 
 @PluginDescriptor(
-	name = "Slayer",
-	description = "Show additional slayer task related information",
-	tags = {"combat", "notifications", "overlay", "tasks"}
+		name = "Slayer",
+		description = "Show additional slayer task related information",
+		tags = {"combat", "notifications", "overlay", "tasks"}
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Although the IDE does 2 tabs automatically for annotations, we prefer one. Also avoid changing whitespacing unless the code was changed.

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 3]
```diff
@@ -168,8 +166,8 @@ protected void startUp() throws Exception
 		overlayManager.add(targetMinimapOverlay);
 
 		if (client.getGameState() == GameState.LOGGED_IN
-			&& config.amount() != -1
-			&& !config.taskName().isEmpty())
+				&& config.amount() != -1
+				&& !config.taskName().isEmpty())
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Don't change whitespacing unless the code itself was also changed. (even though this style is preferred)

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 4]
```diff
@@ -208,8 +206,8 @@ public void onGameStateChange(GameStateChanged event)
 				break;
 			case LOGGED_IN:
 				if (config.amount() != -1
-					&& !config.taskName().isEmpty()
-					&& loginFlag)
+						&& !config.taskName().isEmpty()
+						&& loginFlag)
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Same here

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 5]
```diff
@@ -474,6 +483,59 @@ private void killedOne()
 		infoTimer = Instant.now();
 	}
 
+	private boolean isTarget(NPC npc)
+	{
+		NPCComposition composition = npc.getTransformedComposition();
+
+		if (composition == null || composition.getName() == null)
+			return false;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Add braces

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 6]
```diff
@@ -474,6 +483,59 @@ private void killedOne()
...
+	}
+
+	private void rebuildTargetNames()
+	{
+		targetNames.clear();
+
+		//update the list of target names
+		Task task = Task.getTask(taskName);
+		if (task == null)
+			return;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Add braces

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 7]
```diff
@@ -501,7 +564,7 @@ private void addCounter()
 		BufferedImage taskImg = itemManager.getImage(itemSpriteId);
 		counter = new TaskCounter(taskImg, this, amount);
 		counter.setTooltip(String.format("<col=ff7700>%s</br><col=ffff00>Pts:</col> %s</br><col=ffff00>Streak:</col> %s",
-			capsString(taskName), points, streak));
+				capsString(taskName), points, streak));
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Unnecessary whitespace change

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 8]
```diff
@@ -474,6 +483,59 @@ private void killedOne()
...
+		targetNames.clear();
+
+		//update the list of target names
+		Task task = Task.getTask(taskName);
+		if (task == null)
+			return;
+
+		targetNames.addAll(Arrays.asList(task.getTargetNames()));
+		targetNames.add(taskName.replaceAll("s$", ""));
+
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Remove empty new line

**Comment by AWarbear** [comment]
The code should now match the conventions

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 9]
```diff
@@ -30,10 +30,7 @@
 import java.awt.image.BufferedImage;
 import java.time.Duration;
 import java.time.Instant;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Still has wildcard import ;)

Comment (AWarbear) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: wow I'm blind, sorry

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 10]
```diff
@@ -474,14 +484,66 @@ private void killedOne()
...
-		taskName = name.toLowerCase();
+		taskName = name;
 		amount = amt;
 		save();
 		removeCounter();
 		addCounter();
 		infoTimer = Instant.now();
+
+		Task task = Task.getTask(name);
+		if (task != null)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: This should account for having highlighting enabled or not

**Comment by AWarbear** [comment]
The tests want lowercase task names. Should I and can I change the tests since Adam changed the names to be uppercase?
